doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,339 +823,538 @@ tactic
+
 Application
 ~~~~~~~~~~~
 
 An application is an expression of the following form:
 
-.. tacn:: @qualid {+ @tacarg}
+.. tacn:: @qualid {* @tactic_arg }
 
    The reference :n:`@qualid` must be bound to some defined tactic definition
-   expecting at least as many arguments as the provided :n:`tacarg`. The
-   expressions :n:`@ltac_expr__i` are evaluated to :n:`v__i`, for i = 1, ..., n.
+   expecting at least as many arguments as the provided :n:`@tactic_arg`. The
+   expressions :n:`@tactic_arg__i` are evaluated to :n:`v__i`, for i = 1, ..., n.
 
    .. what expressions ??
 
-Function construction
-~~~~~~~~~~~~~~~~~~~~~
+Function construction: fun
+~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 A parameterized tactic can be built anonymously (without resorting to
 local definitions) with:
 
-.. tacn:: fun {+ @ident} => @ltac_expr
+.. tacn:: fun {+ @name } => @ltac_expr
 
    Indeed, local definitions of functions are a syntactic sugar for binding
    a :n:`fun` tactic to an identifier.
 
-Pattern matching on terms
-~~~~~~~~~~~~~~~~~~~~~~~~~
+   :tacn:`fun` is a :token:`binder_tactic`.
+
+Pattern matching on terms: match
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tacn:: @match_key @ltac_expr__term with {? %| } {+| @match_pattern => @ltac_expr } end
+   :name: lazymatch; match; multimatch
+
+   .. insertprodn match_key cpattern
+
+   .. prodn::
+      match_key ::= lazymatch
+      | match
+      | multimatch
+      match_pattern ::= @cpattern
+      | context {? @ident } [ @cpattern ]
+      cpattern ::= @term
+
+   .. todo I don't fully understand these descriptions.  I think some concepts
+      such as backtracking need to be explained first and that the lanugage
+      can probably be simplified.
+
+   :tacn:`lazymatch`, :tacn:`match` and :tacn:`multimatch` are :token:`ltac_expr1`\s.
 
-We can carry out pattern matching on terms with:
+   .. todo: below: "yield" or "be"??
 
-.. tacn:: match @ltac_expr with {+| @cpattern__i => @ltac_expr__i} end
+   Evaluates :n:`@ltac_expr__term`, which must yield a term, and matches it
+   sequentially with the :token:`match_pattern`\s, which may have
+   metavariables.  When a match is found, metavariable values are substituted
+   into :n:`@ltac_expr`, which is then applied.
 
-   The expression :n:`@ltac_expr` is evaluated and should yield a term which is
-   matched against :n:`cpattern__1`. The matching is non-linear: if a
-   metavariable occurs more than once, it should match the same expression
-   every time. It is first-order except on the variables of the form :n:`@?id`
-   that occur in head position of an application. For these variables, the
+   Matching may continue depending on whether  `lazymatch`, `match` or `multimatch`
+   is specified.
+
+   In the :token:`match_pattern`\s, metavariable values have the form :n:`?@ident`, whereas
+   in the :n:`@ltac_expr`\s, the question mark is omitted.  Choose your metavariable
+   names with care.
+
+   Matching is done with non-linear unification: if a
```

Comment:
Careful.  It's not full-blown unification, as the following fails:
```coq
Goal True /\ (fun x => x) True.
  match goal with
  | [ |- ?x /\ ?x ] => idtac
  end.
```
But I think it's not quite syntactic matching either.  @ppedrot might have an example of these dragons?

Reply:
Wouldn't unifying `True` and `(fun x => x) True` make it second-order unification?  I wouldn't expect unification to do that.  Of course, we have no documentation of how we do unification.

Reply:
Perhaps
```suggestion
   Matching is done with non-linear syntactic unification: if a
```
Note that `let x := constr:(True) in let y := constr:((fun x => x) True) in unify x y` succeeds, while replacing `unify` with `constr_eq` causes it to fail.  My point was about it using something more like `constr_eq` rather than `unify`.

Reply:
I'll revert to the original "Matching is non-linear" and leave a "todo" in the file.

Syntactic matching would be comparing the ASTs, right?  Would `((x + y))` match `1 + 2`?Though maybe it's a silly pattern.

Reply:
> Syntactic matching would be comparing the ASTs, right? 

Yes.

> Would `((x + y))` match `1 + 2`?Though maybe it's a silly pattern.

`((x +y))` does not match`1 + 2`, but `((?x + ?y))` (which parses to `Nat.add ?x ?y`) does match `1 + 2` (which is `Nat.add 1 2`).  Note that matching happens after notations are interpreted (they happen on, basically, the `Set Printing All` version of the terms).